src/Lykke.Job.TransactionHandler/Sagas/Services/ContextFactory.cs

```diff
@@ -0,0 +1,102 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Lykke.Job.TransactionHandler.Queues.Models;
+using Lykke.Service.Assets.Client;
+using Lykke.Service.ClientAccount.Client;
+
+namespace Lykke.Job.TransactionHandler.Sagas.Services
+{
+    public enum OperationStatus
+    {
+        Matched
+    }
+
+    public class ContextFactory : IContextFactory
+    {
+        private readonly IClientTradesFactory _clientTradesFactory;
+        private readonly IClientAccountClient _clientAccountClient;
+        private readonly IAssetsServiceWithCache _assetsServiceWithCache;
+
+        public ContextFactory(
+            IClientTradesFactory clientTradesFactory,
+            IClientAccountClient clientAccountClient,
+            IAssetsServiceWithCache assetsServiceWithCache)
+        {
+            _clientTradesFactory = clientTradesFactory;
+            _clientAccountClient = clientAccountClient;
+            _assetsServiceWithCache = assetsServiceWithCache;
+        }
+
+        public async Task<SwapOffchainContextData> FillTradeContext(SwapOffchainContextData context, TradeQueueItem.MarketOrder order, List<TradeQueueItem.TradeInfo> trades, string clientId)
+        {
+            var marketVolume = trades.Sum(x => x.MarketVolume);
+            var limitVolume = trades.Sum(x => x.LimitVolume);
+
+            context.Order = order;
+            context.Trades = trades;
+            context.ClientTrades = await _clientTradesFactory.Create(order.Id, clientId, trades[0], marketVolume, limitVolume);
+            context.IsTrustedClient = (await _clientAccountClient.IsTrustedAsync(clientId)).Value;
+            context.Status = OperationStatus.Matched;
+
+            if (!context.IsTrustedClient)
+            {
+                var aggregatedTrades = await AggregateSwaps(trades);
+
+                context.SellTransfer = aggregatedTrades.sellTransfer;
+                context.BuyTransfer = aggregatedTrades.buyTransfer;
+
+                var operations = new[] { aggregatedTrades.sellTransfer, aggregatedTrades.buyTransfer };
+
+                foreach (var operation in operations)
+                {
+                    var trade = context.ClientTrades.FirstOrDefault(x =>
```

Comment:
Is field ClientTrades contains only active client trades? This linear search looks not scalable

Reply:
This collection contains just 2 elements.